### PR TITLE
feat: add msgspec support

### PR DESCRIPTION
**Pull request by aarcex3**
# Summary

Added msgspec support as (de)serializer.

# Related Issue
#13

### CODE REVIEW

**Comment by aarcex3** [comment]
Quick question, I saw that you have a `BaseSerializer`  and a `BaseDeserializer` class, but only use them once. I mean, the other (de)serializers don't inherent from thoses classes. Why?

Wouldn't make sense to do something like this?
```python
class MsgspecSerializer(BaseSerializer):
    """
    Convert from dict to msgspec.Struct.
    """

    @classmethod
    def can_apply(cls, klass: type[FromDictSerializableT]) -> bool:
        return msgspec_installed and issubclass(klass, msgspec.Struct)

    @classmethod
    def from_dict(
        cls, klass: type[FromDictSerializableT], values: dict
    ) -> FromDictSerializableT:
        try:
            return msgspec.convert(values, klass)
        except msgspec.ValidationError as e:
            raise DictSerializationError(expected_type=klass.__name__) from e

```

**Comment by martinn** [comment]
Hey @aarcex3 - thanks so much for this!

I will have a more detailed look later today but outside of the failing checks, it's looking very good!

> Quick question, I saw that you have a `BaseSerializer` and a `BaseDeserializer` class, but only use them once. I mean, the other (de)serializers don't inherent from thoses classes. Why?

This is using [Protocol](https://realpython.com/python-protocol/), which allows us to essentially say, any object which behaves in this way (i.e. has the same methods and attributes as `BaseSerializer` or `BaseDeserializer`) can be used as a (de)serializer, it doesn't need to inherit from them. It provides a bit more flexibility as those objects can be anything, and so long as they have that behavior, they can work fine.

**Comment by aarcex3** [comment]
Didn't know that. Thanks for clarifying.

File: tests/test_client_msgspec.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+class PostMsgspecApi(quickapi.BaseApi[ResponseBody]):
+    url = "https://example.com/facts"
+    method = quickapi.BaseHttpMethod.POST
+    request_params = RequestParams
+    request_body = RequestBody
+    response_body = ResponseBody
+    response_errors = {401: ResponseError401}  # noqa: RUF012
+
+
+class TestGetPydanticApi:
```
Comment (martinn) [inline tests/test_client_msgspec.py]:
```suggestion
class TestPostMsgspecApi:
```

File: quickapi/serializers/msgspec.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        def convert_nested(data: dict[str, Any]) -> dict[Any, Any] | None:
+            if isinstance(data, list):
+                return [convert_nested(item) for item in data]
+            elif isinstance(data, dict):
+                return {key: convert_nested(value) for key, value in data.items()}
+            elif isinstance(data, msgspec.Struct):
+                return cls.to_dict(data)
+            return data
+
+        return convert_nested(data)
```
Comment (martinn) [inline quickapi/serializers/msgspec.py]: I'm not familiar with `msgspec`, but wonder what's the difference between `structs.asdict()` and [to_builtins](https://jcristharif.com/msgspec/api.html#msgspec.to_builtins)?

**Review by martinn — COMMENTED** [review]
I'm not sure yet why the pipeline checks seem to be failing on this. It's like it can't find the `msgspec` package but I can see it installed.

Otherwise looks good to me. Just had one tiny change and a question for you.

File: pyproject.toml [hunk 3]
```diff
@@ -42,6 +43,7 @@ markdown-checklist = "^0.4.4"
 requests = ["requests"]
 attrs = ["attrs", "cattrs"]
 pydantic = ["pydantic"]
+msgspec = ["msgspec"]
```
Comment (martinn) [inline pyproject.toml:47]: Need to also add the package under `[tool.poetry.group.dev.dependencies]`.

**Review by martinn — COMMENTED** [review]
This should fix the pipeline checks.

**Comment by martinn** [comment]
If you wouldn't mind updating the main `README.md` to mention that it now supports `msgspec` (and perhaps adding an example after `pydantic`, that would be great please.

Thanks again for doing this.

File: quickapi/serializers/msgspec.py (hunk 2 above)
Comment (aarcex3) [reply quickapi/serializers/msgspec.py]: Now that you mention it, `struct.asdict()` returns a simple dict representation of the Struct, without deep conversion of nested structures.
I changed it to `to_builtins`, which does exactly that.

**Review by martinn — APPROVED** [review]
Looks great. Thanks again for your contribution.

**Pull request merged**